tidy3d/plugins/smatrix/smatrix.py

```diff
@@ -0,0 +1,267 @@
+"""Tools for generating an S matrix automatically from tidy3d simulation and port definitions."""
+
+from typing import List, Tuple, Dict, Any
+import os
+
+import pydantic as pd
+import numpy as np
+
+from ... import Simulation, Box, ModeSpec, ModeMonitor, ModeSource, GaussianPulse, SimulationData
+from ...constants import HERTZ, C_0
+from ...components.types import Direction, Ax
+from ...components.viz import add_ax_if_none, equal_aspect
+
+# fwidth of gaussian pulse in units of central frequency
+FWIDTH_FRAC = 1.0 / 10
+DEFAULT_DATA_DIR = "data"
+
+
+class Port(Box):
+    """Specifies a port in the scattering matrix."""
+
+    direction: Direction = pd.Field(
+        ...,
+        title="Direction",
+        description="'+' or '-', defining which direction is considered 'input'.",
+    )
+    mode_spec: ModeSpec = pd.Field(
+        ModeSpec(),
+        title="Mode Specification",
+        description="Specifies how the mode solver will solve for the modes of the port.",
+    )
+    mode_indices: Tuple[pd.NonNegativeInt] = pd.Field(
```

Comment:
Maybe this has more to do with the mode solver: what happens if mode indices are requested for a mode not supported by the structure? E.g., TM10 for a rectangular waveguide.

Reply:
this is still a mode solver issue.  My understanding is that the mode solver will "find" some modes because they exist numerically, but they will not be physical.  Our plan is to eventually filter out these unphysical modes but I'm not sure how to add this as a check in the model so am inclined to just trust the user knows what they are doing if they set this.  Note that the validator for this field checks whether `mode_spec.num_modes` is large enough to support these indices.